relay-general/src/protocol/contexts/browser.rs

```diff
@@ -21,3 +24,229 @@ impl BrowserContext {
+        let browser = Browser::from_ua_client_hint(client_hints.sec_ch_ua?)?;
+        let version = client_hints.sec_ch_ua_full_version?.to_owned();
+
+        Some(Self {
+            name: Annotated::new(browser.to_string()),
+            version: Annotated::new(version),
+            ..Default::default()
+        })
+    }
+
+    fn from_user_agent(user_agent: &str) -> Option<Self> {
+        let browser = user_agent::parse_user_agent(user_agent);
+
+        if !store::user_agent::is_known(&browser.family) {
+            return None;
+        }
+
+        Some(Self {
+            name: Annotated::from(browser.family),
+            version: Annotated::from(store::user_agent::get_version(
+                &browser.major,
+                &browser.minor,
+                &browser.patch,
+            )),
+            ..BrowserContext::default()
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum Browser {
+    Chrome,
+    Edge,
+    Safari,
+    Brave,
+    Opera,
+    Firefox,
+    Other(String),
+}
+
+impl std::convert::From<String> for Browser {
+    fn from(s: String) -> Self {
+        match s.as_str() {
+            "Google Chrome" => Browser::Chrome,
+            "Microsoft Edge" => Browser::Edge,
+            "Firefox" => Browser::Firefox,
+            "Safari" => Browser::Safari,
+            "Brave" => Browser::Brave,
+            "Opera" => Browser::Opera,
+            _ => Browser::Other(s.to_string()),
+        }
+    }
+}
+
+impl std::string::ToString for Browser {
+    fn to_string(&self) -> String {
+        match self {
+            Browser::Chrome => "Google Chrome".to_string(),
+            Browser::Edge => "Microsoft Edge".to_string(),
+            Browser::Firefox => "Firefox".to_string(),
+            Browser::Safari => "Safari".to_string(),
+            Browser::Brave => "Brave".to_string(),
+            Browser::Opera => "Opera".to_string(),
+            Browser::Other(s) => s.clone(),
+        }
+    }
+}
+
+impl Browser {
+    /// the sec-ch-ua field looks something like this:
+    /// "Not_A Brand";v="99", "Google Chrome";v="109", "Chromium";v="109"
+    ///
+    /// this function attempts to detect the field with the browser, parse it as a browser enum and
+    /// return that. if it fails to parse it as a browser, it uses the "other" variant of browser enum
+    ///
+    /// returns None if no browser field detected
+    fn from_ua_client_hint(s: &str) -> Option<Self> {
+        for item in s.split(',') {
+            // if it contains one of these then we can know it isn't a browser field. atm chromium
+            // browsers are the only ones supporting client hints.
+            if item.contains("Brand")
+            || item.contains("Chromium")
+            || item.contains("Gecko") // useless until firefox and safari support client hints
+            || item.contains("WebKit")
+            {
+                continue;
+            }
```

Comment:
By only skipping known non-browsers, we risk accepting an unknown non-browser as browser. I would go the other way around, and only return something from this function if the browser is part of the `Browser` enum.

Then you could easily split on `;` first, to avoid the use of `contains`.